zjit/src/hir.rs

```diff
@@ -3258,79 +3309,61 @@ pub fn iseq_to_hir(iseq: *const rb_iseq_t) -> Result<Function, ParseError> {
     let mut fun = Function::new(iseq);
 
     // Compute a map of PC->Block by finding jump targets
-    let BytecodeInfo { jump_targets, has_blockiseq } = compute_bytecode_info(iseq);
+    let jit_entry_insns = jit_entry_insns(iseq);
+    let BytecodeInfo { jump_targets, has_blockiseq } = compute_bytecode_info(iseq, &jit_entry_insns);
     let mut insn_idx_to_block = HashMap::new();
     for insn_idx in jump_targets {
+        // Prepend a JIT entry block if it's a jit_entry_insn.
+        // compile_entry_block() assumes that a JIT entry block jumps to the next block.
+        if jit_entry_insns.contains(&insn_idx) {
+            let jit_entry_block = fun.new_block(insn_idx);
+            fun.jit_entry_blocks.push(jit_entry_block);
+        }
         insn_idx_to_block.insert(insn_idx, fun.new_block(insn_idx));
     }
 
     // Check if the EP is escaped for the ISEQ from the beginning. We give up
     // optimizing locals in that case because they're shared with other frames.
     let ep_escaped = iseq_escapes_ep(iseq);
 
-    // Compile an interpreter entry to the first block
-    let first_block = insn_idx_to_block[&0];
-    compile_entry_block(fun.entry_block, first_block, &mut fun, false);
 
-    // Compile a JIT entry to the first block
-    compile_entry_block(fun.jit_entry_block, first_block, &mut fun, true);
+    // Compile an entry_block for the interpreter
+    compile_entry_block(&mut fun, &jit_entry_insns);
 
     // Iteratively fill out basic blocks using a queue
-    // TODO(max): Basic block arguments at edges
```

Comment:
Don't think you meant to delete this TODO.

Reply:
added it back https://github.com/ruby/ruby/pull/14653/commits/23546770af4f3e9ce50730f17b78e80ea64ef856